### PR TITLE
Support secondary directory for log writing.

### PR DESCRIPTION
**Pull request by LykxSassinator**
## Brief Introduction
This pr is used to support the `secondary directory` feature, referred to ISSUE: #257.
It contains:
* Add an extra configuration `sub-dir` to enable this feature. Default is `None`, and if set with `Some(...)`, this directory will be used if the main dir is full, specified by `dir` option.
* Add an extra class, named as `StorageInfo` in `pipe.rs`, to make this feature compatible to other existing features, i.e. recycle logs, recoverying and so on.

### CODE REVIEW

**Comment by LykxSassinator** [comment]
It's a prototype for closing #257, and all reasonable suggestions are acceptable.

**Comment by LykxSassinator** [comment]
It's a prototype for closing #257, and all reasonable suggestions are acceptable.

**Comment by LykxSassinator** [comment]
@tabokie Please take a review, this is a prototype for implementing `secondary dir`. If this design is a bit confusing, we can briefly make a offline discussion later.

**Comment by LykxSassinator** [comment]
Please hold on, the design of it should be refined and polished, to simplify the whole structure.

**Comment by LykxSassinator** [comment]
Please hold on, the design of it should be refined and polished, to simplify the whole structure.

File: src/engine.rs [hunk 1]
```diff
@@ -191,7 +191,18 @@ where
             debug_assert_eq!(writer.perf_context_diff.write_wait_duration, Duration::ZERO);
             perf_context += &writer.perf_context_diff;
             set_perf_context(perf_context);
-            writer.finish()?
+            // Retry if `writer.finish()` returns a special err, remarking there still
+            // exists free space for this `LogBatch`.
+            let ret = writer.finish();
```
Comment (tabokie) [inline src/engine.rs:196]: This is not enough. fsync failure will panic. You should let leader do the retry.

**Comment by tabokie** [comment]
I think letting leader do all the retrying is simpler. When a write fails with NOSPC, the leader will close the current active file (it internally truncates un-sync-ed parts), and create a new file and retry the whole write group.

**Comment by LykxSassinator** [comment]
I hold a different view on it.
As I do not expect the current writing group hanging on one LogBatch writing because of `NOSPC` err, I think the current leader should only be responsible for successfully written `LogBatch`s, and the followers, failed with `NOSPC`, should put them LogBatch into next writing group and wait for the next leader to do the `append` ops. And that's the core reason why I implemented the `retry` strategy.

**Comment by LykxSassinator** [comment]
I hold a different view on it.
As I do not expect the current writing group hanging on one LogBatch writing because of `NOSPC` err, I think the current leader should only be responsible for successfully written `LogBatch`s, and the followers, failed with `NOSPC`, should put them LogBatch into next writing group and wait for the next leader to do the `append` ops. And that's the core reason why I implemented the `retry` strategy.

File: src/engine.rs [hunk 2]
```diff
@@ -164,12 +174,35 @@ where
...
+                if force_rotate {
+                    if let Err(e) = self.pipe_log.rotate(LogQueue::Append) {
+                        panic!(
+                            "Cannot rotate {:?} queue due to IO error: {}",
+                            LogQueue::Append,
+                            e
+                        );
+                    }
+                }
                 if let Err(e) = self.pipe_log.maybe_sync(LogQueue::Append, sync) {
```
Comment (tabokie) [inline src/engine.rs:206]: It's still incorrect. The NOSPC error would happen here. You need to check its error as well.

File: src/engine.rs [hunk 3]
```diff
@@ -191,7 +224,17 @@ where
...
-            writer.finish()?
+            // Retry if `writer.finish()` returns a special err, remarking there still
+            // exists free space for this `LogBatch`.
+            let ret = writer.finish();
+            if let Err(Error::Other(_)) = ret {
+                log_batch.reset_to_encoded_state();
+                // Here, we will retry this LogBatch `append` by appending this writer
+                // to the next write group, and the current write leader will not hang
+                // on this write and will return timely.
+                return self.write(log_batch, sync);
```
Comment (tabokie) [inline src/engine.rs:235]: This will probably enter infinite recursive calls if the disk is still full after rotate.

File: src/engine.rs [hunk 4]
```diff
@@ -148,9 +148,19 @@ where
                 let now = Instant::now();
                 let _t = StopWatch::new_with(&*ENGINE_WRITE_LEADER_DURATION_HISTOGRAM, now);
                 let file_context = self.pipe_log.fetch_active_file(LogQueue::Append);
+                // Flag on whether force to flush the current active file or not.
+                let mut force_rotate = false;
```
Comment (tabokie) [inline src/engine.rs:152]: Two things here, (1) I don't want the engine to care about "rotate" details. We can at most implement a `fn retriable` for the error which is true for disk full (2) We need to assume all buffered data is lost if `fsync` fails. To tolerate disk full we need a way to detect "if there's any buffered write from last write group". If so we must panic regardless of whether it's a disk full error or not.

I haven't thought of an elegant abstraction to handle both of these cases, maybe we can rethink it next week.

File: src/file_pipe_log/pipe_builder.rs [hunk 5]
```diff
@@ -99,108 +101,201 @@ impl<F: FileSystem> DualPipesBuilder<F> {
     /// Scans for all log files under the working directory. The directory will
     /// be created if not exists.
     pub fn scan(&mut self) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply src/file_pipe_log/pipe_builder.rs:103]: The modifications are not change, just spliting the procedure of `scan` into three parts. It's not approapriate to code it with two `fs::read_dir(path)?.for_each`s, as the definition of `cfg.secondary-dir` is `Option<String>`.

File: src/engine.rs (hunk 3 above)
Comment (LykxSassinator) [reply src/engine.rs:235]: Just like your concerns in the following comments `Two things here ...`, I still wonder whether there exists other elegant ways to handle it.

File: src/engine.rs (hunk 4 above)
Comment (LykxSassinator) [reply src/engine.rs:152]: About the second suggestion u mentioned above, I tried to figure out whether there exists an elegant solution to this corner case recently, but ended with failure. **I think it's hard to detect "if there's any buffered write from last write group".**

And I hold the view that if `fsync` is failed, there should be some hardware errs raised by IO device.
Refer to **https://wiki.postgresql.org/wiki/Fsync_Errors**, I think we also can take the same processing if `fsync` returns a unexpected error, **`panic`** is enough for it.

So, current `maybe_sync` is enought for `nospc` err:
- If `rotate` fails, it means:
  - There might have no free space or no stale files to reuse, should `panic`.
  - `fsync` is failed, as the previous annotations show, should `panic`.
- If `fsync` fails, just `panic`.

**Comment by LykxSassinator** [comment]
PTAL cc @tabokie , several modifications have been supplied according to previous suggestions.

File: src/engine.rs (hunk 4 above)
Comment (tabokie) [reply src/engine.rs:152]: The point is `fsync` can fail even if `pwrite` succeeds. But I agree in a practical sense this is unlikely to happen.

File: src/engine.rs (hunk 3 above)
Comment (tabokie) [reply src/engine.rs]: This issue is not addressed.

File: src/log_batch.rs [hunk 6]
```diff
@@ -555,6 +555,14 @@ enum BufState {
     /// # Invariants
     /// LOG_BATCH_HEADER_LEN <= buf.len()
     Sealed(usize, usize),
```
Comment (tabokie) [inline src/log_batch.rs:557]: Let's not add yet another state. Change `Sealed` to contain a third field "original checksum". During `prepare_write`, if the state is `Sealed`, we can use this field to re-sign checksum.

File: src/log_batch.rs [hunk 7]
```diff
@@ -779,14 +790,17 @@ impl LogBatch {
...
+    /// Makes preparations for the write of `LogBatch`.
     #[inline]
     pub(crate) fn prepare_write(&mut self, file_context: &LogFileContext) -> Result<()> {
         match self.buf_state {
             BufState::Encoded(header_offset, entries_len) => {
                 LogItemBatch::prepare_write(&mut self.buf, file_context)?;
                 self.buf_state = BufState::Sealed(header_offset, entries_len);
             }
+            BufState::ReSealing(_, _) => {
+                LogItemBatch::prepare_write(&mut self.buf, file_context)?;
```
Comment (tabokie) [inline src/log_batch.rs:802]: This doesn't make sense, the signature will be `checksum^first_file^second_file`. I wonder why the tests didn't catch it.

File: src/file_pipe_log/pipe_builder.rs [hunk 8]
```diff
@@ -99,108 +101,203 @@ impl<F: FileSystem> DualPipesBuilder<F> {
...
+            if let Some(secondary_dir) = self.cfg.secondary_dir.as_ref();
+            if let Ok(0) = parse_file_range(
+                    secondary_dir,
+                    StorageDirType::Secondary,
+                    &mut append_id_range,
+                    &mut rewrite_id_range,
+                    &mut append_file_dict,
+                    &mut rewrite_file_dict,
+                );
+            if let Ok(true) = from_same_dev(&self.cfg.dir, secondary_dir);
```
Comment (tabokie) [inline src/file_pipe_log/pipe_builder.rs:238]: This is wrong, user would've migrated the directory from another disk. Let's check the device during config validation instead.

File: src/log_batch.rs (hunk 6 above)
Comment (LykxSassinator) [reply src/log_batch.rs:557]: `ReSealing` is a special state to handle the recursive `write` problem, remarking that the LogBatch has entered the `rewrite` state. 

With this special state, the log_batch will follow the state transformating path `Write(Open -> Encoded -> Sealed)` => `Rewrite(Sealed -> Resealing -> end with panic || end with successful "sync")`

So, if `rewrite` failed,  the next `return self.write(log_batch, sync)` would `panic` in the next checking of `log_batch.prepare_rewrite()`.

File: src/file_pipe_log/pipe_builder.rs (hunk 8 above)
Comment (LykxSassinator) [reply src/file_pipe_log/pipe_builder.rs:238]: But I prefer doing dir checking here.

Here, only when "the `secondary-dir` had no valid log files" and "`from_same_dev` returned `true`", we could ignore the `secondary-dir`. And whether the migrated directory owns valid files has been checked by `if let Ok(0) = parse_file_range(...)`.

If we made dir checking during the config validation, we would not know _whether there exists valid log files in the `secondary-dir` when `from_same_dev(...)` returns `true`_. In this implementation,  we should check it again when scanning the two dir and make the decision whether we could ignore the `secondary-dir`.

Comment (tabokie) [reply src/file_pipe_log/pipe_builder.rs:238]: So what's the point of doing the check at all? We already allow a secondary directory on the same device in the case of "user migration". The way I see it, you either deny it or accept it, there shouldn't be any middle ground.

File: src/log_batch.rs (hunk 6 above)
Comment (tabokie) [reply src/log_batch.rs:557]: It's too obscure to use this as a recursion limit. Using a loop to retry is more explicit and understandable.

File: src/file_pipe_log/pipe_builder.rs (hunk 8 above)
Comment (LykxSassinator) [reply src/file_pipe_log/pipe_builder.rs:238]: No, we don't allow user use a `secondary-dir` on the same device, the strategy here is to check and ignore this case, by `reset cfg.secondary-dir = None` and logging with warning `"sub-dir ({}) and dir ({}) are on same device, ignore it"`.

Maybe I should add an extra specification to `secondary-dir`, which is only valid when it's on different device?

Comment (tabokie) [reply src/file_pipe_log/pipe_builder.rs:238]: > Here, only when "the secondary-dir had no valid log files" and "from_same_dev returned true", we could ignore the secondary-dir.

You said it before, this case is tolerated, i.e. the secondary directory on the same device can still be used.

Comment (LykxSassinator) [reply src/file_pipe_log/pipe_builder.rs:238]: Emmm.....more precisely, I expect and strongly recommend that users should use different dir on different devices respectively.

However, as u mentioned above, some users might migrate the `secondary-dir` from another device to the device, `dir` belonging to. And in this case, I think we should tolerate it, considering data integrity. And that's why I introduce this check here.

Comment (tabokie) [reply src/file_pipe_log/pipe_builder.rs:238]: My point is, if the code can still work (and is expected to work) if the secondary-dir is located on the same device, then there's no need to impose a restriction. By banning this case, you are in fact making it more difficult to test. That's why I said "you either deny it or accept it, there shouldn't be any middle ground".

Of course, you can still emit a warning (preferably during config validation, you can call `read_dir` there to detect if the directory is empty).

**Comment by LykxSassinator** [comment]
After merging #269, activate this pr again.

Ping @tabokie .

File: src/file_pipe_log/pipe_builder.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tabokie) [reply src/file_pipe_log/pipe_builder.rs:110]: The change is still too big and I don't see a clear purpose. You can keep the `scan_dir` simple and short as before, and do the file handle initialization early, then sort the `FileToRecover` list afterwards.

File: src/file_pipe_log/pipe.rs [hunk 9]
```diff
@@ -22,10 +23,91 @@ use crate::{perf_context, Error, Result};
...
+#[repr(u8)]
+#[derive(Clone, Copy, PartialEq, Eq, Debug, EnumIter)]
+pub enum StorageDirType {
+    Main = 0,
+    Secondary = 1,
+}
+
+/// Represents the info of storage dirs, including `main dir` and
+/// `secondary dir`.
+struct StorageInfo {
```
Comment (tabokie) [inline src/file_pipe_log/pipe.rs:35]: StorageManager, and use path_id to refer to path instead of storage_type.

File: src/file_pipe_log/pipe.rs [hunk 10]
```diff
@@ -242,9 +349,7 @@ impl<F: FileSystem> SinglePipe<F> {
     /// Synchronizes all metadatas associated with the working directory to the
     /// filesystem.
     fn sync_dir(&self) -> Result<()> {
```
Comment (tabokie) [inline src/file_pipe_log/pipe.rs:351]: This is a heavy operation. Let's be specific about which directory to sync.

File: src/engine.rs [hunk 11]
```diff
@@ -139,51 +139,111 @@ where
         let start = Instant::now();
         let len = log_batch.finish_populate(self.cfg.batch_compression_threshold.0 as usize)?;
         let block_handle = {
```
Comment (tabokie) [inline src/engine.rs:141]: How about
```rust
let block_handle = loop {
  attempt += 1;
  let res = do_some_writes();
  if res.is_ok() || attempt >= MAX_RETRY {
    break res;
  }
}
```

File: src/engine.rs [hunk 12]
```diff
@@ -139,51 +139,111 @@ where
...
+                            ))));
+                            continue;
+                        }
+                        writer.set_output(res);
+                    }
+                    perf_context!(log_write_duration).observe_since(now);
+                    if force_rotate {
+                        // If the leader failed to `rotate` a new log for the un-synced LogBatches,
+                        // it means that it encounters unexpected errors, and should panic.
+                        if let Err(e) = self.pipe_log.rotate(LogQueue::Append) {
```
Comment (tabokie) [inline src/engine.rs:205]: Put the rotate inside `append`, which is the whole point of #269.

File: src/file_pipe_log/pipe_builder.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply src/file_pipe_log/pipe_builder.rs:110]: On my view, I think it's more clear than before.

`scan` has been splited into the following steps:
1. scanning dir to get file_seq_range of append logs and rewrite logs
  + `scan_dir` in **Main dir**;
  + `scan_dir` in **Secondary dir**, if it had been specified by `cfg.secondary-dir`;
2. Check and clear stale metadata of logs
  + `clear_stale_metadata` in **Main dir**;
  + `clear_stale_metadata` in **Secondary dir**, if it had been specified by `cfg.secondary-dir`;
3. Build file_list vector;

Still confusing?

[the remark this answers is not in the document]
Comment (tabokie) [reply src/file_pipe_log/pipe_builder.rs:110]: There're two things here, one is to keep diff as short as possible. I do refactor all the time, but they are often centered around a change of abstraction (i.e. interaction between different modules, usually involves changing function interface and data types). I don't often refactor the internal implementation unless it significantly improves readability.

Then let's go to the code itself, is the readability improved? IMO it is not. The LOC increased 100%. And you introduced several types for the sake of refactoring (adding glue types usually indicates a bad abstraction). In particular I don't think the `min_id`/`max_id` approach is suitable anymore.

File: src/file_pipe_log/pipe.rs (hunk 9 above)
Comment (tabokie) [reply src/file_pipe_log/pipe.rs]: path_id as in an integer. You can make the StorageManager manages a vector of directories, and you can also let it take care of directory lock as well. (Pass it to `Pipe::open`).

File: src/file_pipe_log/pipe_builder.rs [hunk 13]
```diff
@@ -90,31 +91,125 @@ impl<F: FileSystem> DualPipesBuilder<F> {
...
+    /// Scans for all log files under the working directory.
     pub fn scan(&mut self) -> Result<()> {
-        let dir = &self.cfg.dir;
-        let path = Path::new(dir);
-        if !path.exists() {
-            info!("Create raft log directory: {}", dir);
-            fs::create_dir(dir)?;
-            self.dir_lock = Some(lock_dir(dir)?);
-            return Ok(());
+        // Scan main `dir` and `secondary-dir`, if `secondary-dir` is valid.
```
Comment (tabokie) [inline src/file_pipe_log/pipe_builder.rs:102]: Let me be more verbose this time:
```rust
struct PipeBuilder {
  // Only available after a successful `scan`.
  dir_manager: Option<Arc<DirectoryManager>>,
}
struct DirectoryManager {
  paths: Vec<PathBuf>,
  locks: Vec<File>,
}

fn scan(&self) {
  // setup directories 
  let mut dirs = DirectoryManager::new();
  dirs.add(self.cfg.dir)?;
  dirs.add(self.second_dir)?;
  for path_id in 0..dirs.len() {
    self.scan_dir(path_id);
  }
  self.rewrite_files.sort();
  self.append_files.sort();
  for queue, files in [] {
    if files.is_empty() { continue; }
    // check consecutiveness
    let mut invalid_files = 0;
    let mut current_seq = files[0].seq;
    for (i, f) in files.enumerate() {
      if f.seq > current_seq {
        warn!("hole");
        current_seq = f.seq + 1;
        invalid_files = i;
      } else if {
        f.seq < current_seq;
        return Error::InvalidArgument("Duplicate file");
      }
    }
    files.drain(..invalid_files);
    if files.is_empty() { continue; }
    // cleanup metadata
    let delete_start = {...}
    'cleanup: for seq in delete_start..files[0].seq {
      for path_id in 0..dirs.len() {
        if self.file_system.exists_metadata(dirs.path(path_id)) {
          if let Err(e) = self.file_system.delete_metadata() { ...; break 'cleanup; }
        }
      }
    }
  }

  self.dir_manager = Arc::new(dirs);
}

fn scan_dir(&self) {
  for f in fs::read_dir() {
    self.rewrite_files.push(...);
  }
}

fn build(self) {
  SinglePipe::new(self.dir_manager.clone(), ...);
}
```

File: src/util.rs [hunk 14]
```diff
@@ -325,6 +325,56 @@ pub trait Factory<Target>: Send + Sync {
...
+    /// Check the given `dir` is valid or not. If `dir` not exists,
+    /// it would be created automatically.
+    pub(crate) fn validate_dir(dir: &str) -> IoResult<()> {
+        let path = Path::new(dir);
+        if !path.exists() {
+            std::fs::create_dir(dir)?;
+            return Ok(());
+        }
+        if !path.is_dir() {
+            return Err(std::io::Error::new(
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply src/util.rs:343]: Not returned with a std::io::Error? Or u means that the `ErrorKind` is not suitable ?

[the remark this answers is not in the document]
Comment (tabokie) [reply src/util.rs:343]: The former. Actually I think it makes more sense to move this function to config.rs.

**Comment by LykxSassinator** [comment]
This pr is based on a too obsolete `master` branch, which needs to be rebase.

So, I've reviewed the previous comment and re-built another pr for tackling this issue.

**Review by tabokie — COMMENTED** [review]
(Some pending comments from months ago)

File: src/log_batch.rs [hunk 15]
```diff
@@ -551,10 +529,10 @@ enum BufState {
     /// state only briefly exists between encoding and writing, user operation
     /// will panic under this state.
     /// # Content
-    /// (header_offset, entries_len)
+    /// (header_offset, entries_len, signature)
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply src/log_batch.rs:532]: This change on `BufState::Sealed()` has been refactored in the new PR.

File: src/log_batch.rs [hunk 16]
```diff
@@ -779,13 +760,35 @@ impl LogBatch {
...
+                self.buf_state = BufState::Sealed(header_offset, entries_len, new_signature);
+            }
+            BufState::Sealed(header_offset, entries_len, old_signature) => {
+                // Re-seal the LogBatch.
+                match (old_signature, new_signature) {
+                    (Some(old), Some(new)) => {
+                        sign_checksum(&mut self.buf, Some(old ^ new))?;
+                    }
+                    (Some(old), None) => {
+                        sign_checksum(&mut self.buf, Some(old))?;
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply src/log_batch.rs:784]: Same as above.

File: src/config.rs [hunk 17]
```diff
@@ -183,10 +214,31 @@ impl Config {
...
+fn validate_dir(dir: &str) -> Result<()> {
+    use std::path::Path;
+
+    let path = Path::new(dir);
+    if !path.exists() {
+        std::fs::create_dir(dir)?;
+        return Ok(());
+    }
+    if !path.is_dir() {
+        return Err(box_err!("Not directory: {}", dir));
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply src/config.rs:228]: Same as above.

File: src/engine.rs [hunk 18]
```diff
@@ -154,14 +163,16 @@ where
...
                     let log_batch = writer.mut_payload();
                     let res = self.pipe_log.append(LogQueue::Append, log_batch);
                     writer.set_output(res);
                 }
                 perf_context!(log_write_duration).observe_since(now);
                 if sync {
-                    // As per trait protocol, this error should be retriable. But we panic anyway to
-                    // save the trouble of propagating it to other group members.
+                    // As per trait protocol, this error should be retriable. But we panic
+                    // anyway to save the trouble of propagating it to
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply src/engine.rs:174]: Tackled in the new pr.

File: src/engine.rs [hunk 19]
```diff
@@ -178,7 +189,27 @@ where
             debug_assert_eq!(writer.perf_context_diff.write_wait_duration, Duration::ZERO);
             perf_context += &writer.perf_context_diff;
             set_perf_context(perf_context);
-            writer.finish()?
+            // Retry if `writer.finish()` returns a special 'Error::Other', remarking that
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply src/engine.rs:192]: Done.

**Pull request closed without merging**